crates/neon/src/types_impl/buffer/types.rs

```diff
@@ -0,0 +1,765 @@
+        // `Context` is in scope. This means that the referenced data is valid *at least* as long
+        // as `Context`, even if the `Handle` is dropped.
+        unsafe { sys::buffer::as_mut_slice(cx.env().to_raw(), self.to_raw()) }
+    }
+
+    fn as_mut_slice<'cx, 'a, C>(&mut self, cx: &'a mut C) -> &'a mut [Self::Item]
+    where
+        C: Context<'cx>,
+    {
+        // # Safety
+        // See `as_slice`
+        unsafe { sys::buffer::as_mut_slice(cx.env().to_raw(), self.to_raw()) }
+    }
+
+    fn try_borrow<'cx, 'a, C>(&self, lock: &'a Lock<C>) -> Result<Ref<'a, Self::Item>, BorrowError>
+    where
+        C: Context<'cx>,
+    {
+        // The borrowed data must be guarded by `Ledger` before returning
+        Ledger::try_borrow(&lock.ledger, unsafe {
+            sys::buffer::as_mut_slice(lock.cx.env().to_raw(), self.to_raw())
+        })
+    }
+
+    fn try_borrow_mut<'cx, 'a, C>(
+        &mut self,
+        lock: &'a Lock<C>,
+    ) -> Result<RefMut<'a, Self::Item>, BorrowError>
+    where
+        C: Context<'cx>,
+    {
+        // The borrowed data must be guarded by `Ledger` before returning
+        Ledger::try_borrow_mut(&lock.ledger, unsafe {
+            sys::buffer::as_mut_slice(lock.cx.env().to_raw(), self.to_raw())
+        })
+    }
+
+    fn byte_length<'cx, C: Context<'cx>>(&self, cx: &mut C) -> usize {
+        unsafe { sys::buffer::size(cx.env().to_raw(), self.to_raw()) }
+    }
+}
+
+/// The standard JS [`ArrayBuffer`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer) type.
+///
+/// # Example
+///
+/// ```
+/// # use neon::prelude::*;
+/// use neon::types::buffer::TypedArray;
+///
+/// fn make_sequence(mut cx: FunctionContext) -> JsResult<JsArrayBuffer> {
+///     let len = cx.argument::<JsNumber>(0)?.value(&mut cx);
+///     let mut buffer = cx.array_buffer(len as usize)?;
+///
+///     for (i, elem) in buffer.as_mut_slice(&mut cx).iter_mut().enumerate() {
+///         *elem = i as u8;
+///     }
+///
+///     Ok(buffer)
+/// }
+/// ```
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct JsArrayBuffer(raw::Local);
+
+impl JsArrayBuffer {
+    /// Constructs a new `JsArrayBuffer` object, safely zero-filled.
+    pub fn new<'a, C: Context<'a>>(cx: &mut C, len: usize) -> JsResult<'a, Self> {
+        let result = unsafe { sys::arraybuffer::new(cx.env().to_raw(), len) };
+
+        if let Ok(buf) = result {
+            Ok(Handle::new_internal(Self(buf)))
+        } else {
+            Err(Throw::new())
+        }
+    }
+
+    /// Construct a new `JsArrayBuffer` from bytes allocated by Rust
+    pub fn external<'a, C, T>(cx: &mut C, data: T) -> Handle<'a, Self>
+    where
+        C: Context<'a>,
+        T: AsMut<[u8]> + Send + 'static,
+    {
+        let env = cx.env().to_raw();
+        let value = unsafe { sys::arraybuffer::new_external(env, data) };
+
+        Handle::new_internal(Self(value))
+    }
+}
+
+impl<'cx> Handle<'cx, JsArrayBuffer> {
```

Comment:
I suggest a few changes to this:
* Renaming to `to_region` for consistency
* Taking `&self` instead of `self`
* Implementing on `JsArrayBuffer` instead of `Handle` for better docs

```rust
impl JsArrayBuffer {
    pub fn to_region<T: Binary>(&self, byte_offset: usize, len: usize) -> Region<T> {
        Region {
            buffer: Handle::new_internal(JsArrayBuffer(self.0)),
            byte_offset,
            len,
            phantom: PhantomData,
        }
    }
}
```

The downside of this approach is that the lifetime is now tied to the borrow instead `'cx`. This means that a `Region` can no longer escape a helper function.

The better documentation *might* be worth the limitation, but we could also fix this up if [`arbitrary_self_types`](https://github.com/rust-lang/rust/issues/44874) stabilizes.

Reply:
Update: We discussed all these in the RFC, and I've updated this PR to match the RFC.